### PR TITLE
"Minecraft Status" Instant Answer

### PR DESCRIPTION
**Pull request by destruc7i0n**
**What does your Instant Answer do?**
It will go and check the status of Minecraft, including Skins, Textures, Session, Login, Realms and their own Website.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It has a very easy an simple to read format that can tell the answer you need quickly, other sites have many different places to look at, this gives an overview.

**What is the data source for your Instant Answer? (Provide a link if possible)**
I am currently getting data from Mojang itself (https://status.mojang.com/check) and also from Mojang's Realms (http://wiki.vg/Realms_API#GET_.2Fmco.2Favailable)
I am currently processing the info and getting the best status on my side.

**Why did you choose this data source?**
I chose these as they are the official creators of the game. I am mirroring it for simplicity sake.

**Are there any other alternative (better) data sources?**
I know of one by someone who has done an online version of this (http://xpaw.ru/mcstatus). When I approached him to see if I could use it he said no. :disappointed:

**What are some example queries that trigger this Instant Answer?**
For example "minecraft status" and "mc status". 

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
This will be useful for gamers (especially Minecraft players!)

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No.

**Which existing Instant Answers will this one supersede/overlap with?**
This will overlap with IsItUp as if I were to type "is mc down" it would go and use IsItUp for that instant answer, not mine.

**Are you having any problems? Do you need our help with anything?**
I used to have problems with design, but I changed it.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
No terms, just that you do not parse me every second :laughing: 

**Where did you hear about DuckDuckHack? (For first time contributors)**
I heard of it when I was working on my own use of the DDG API for a [Twitter bot](https://twitter.com/BotTellMe). I was wanting to know if there was a way to get the status of Minecraft and when there was not I decided to make one myself! 

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
My preview

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [X] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [X] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [X] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [X] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

### CODE REVIEW

**Comment by jagtalon** [comment]
@destruc7i0n can you give us screenshots of the other statuses?

**Comment by destruc7i0n** [comment]
@jagtalon There is no others, I am converting the data on my side to give "good" ,"shaky" or "down" (http://thedestruc7i0n.ca/mcstatusraw)

**Comment by mattr555** [comment]
@destruc7i0n You could definitely use https://status.mojang.com/check as the `spice to` and figure out the display in javascript rather than use the intermediate step of your server (if you're comfortable with that) :sparkles:

**Comment by destruc7i0n** [comment]
@mattr555 I could, but I would be missing realms. Realms is a very big part of MC and it is only at another Endpoint that needs different auth.

**Comment by mattr555** [comment]
@destruc7i0n Ah, ok, thanks for the explanation!

@jagtalon It's cool seeing all of these new status IAs; maybe we could make a new template group for them?

**Comment by destruc7i0n** [comment]
Updated image to have the moreAt

File: t/MinecraftStatus.t [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        '/js/spice/minecraft_status/status',
+        call_type => 'include',
+        caller => 'DDG::Spice::MinecraftStatus'
+    ),
+    'mc status' => test_spice(
+        '/js/spice/minecraft_status/status',
+        call_type => 'include',
+        caller => 'DDG::Spice::MinecraftStatus'
+    ),
+    'mcstatus' => undef;
```
Comment (mattr555) [inline t/MinecraftStatus.t]: @destruc7i0n Please remove the semicolon here, it's causing your [tests to fail](https://travis-ci.org/duckduckgo/zeroclickinfo-spice/jobs/47955345#L1437).

We might want to add one more to trigger "is minecraft up?" to be on the safe side :)

File: lib/DDG/Spice/MinecraftStatus.pm [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+name "Minecraft Status";
+source "help.mojang.com";
+description "Gives the current status of Minecraft";
+primary_example_queries "is minecraft down", "minecraft status";
+category 'special';
+topics 'words_and_games';
+code_url "https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/MinecraftStatus.pm";
+attribution web     => ['https://thedestruc7i0n.ca', "Colin D'Souza"],
+			twitter => ['https://www.twitter.com/TheDestruc7i0n',"Colin D'Souza"],
+    		github  => ['destruc7i0n', "Colin D'Souza"];
```
Comment (mattr555) [inline lib/DDG/Spice/MinecraftStatus.pm]: This is more of a style thing, but could we fix the line indenting in this file? :neckbeard:

**Comment by mattr555** [comment]
@destruc7i0n What kind of additional message will be displayed to the user when everything is not operating normally? Will your service provide insight as to which Minecraft services are down? Just trying to get a feel as to what data we can show the user :rocket:

**Comment by destruc7i0n** [comment]
@mattr555 Hey, I was actually thinking that they would just go to the Mojang website and see that data there, but then there is Realms. They do not show realms status there. That is why I was thinking of a smart way but my lack of knowledge of Handlebars and beginner status of JavaScript blocks this, can I PM you to discuss?

**Comment by mattr555** [comment]
Hey @destruc7i0n, if you want to run the tests on your code before pushing it to github, you can use `duckpan test`. We can probably remove the `is_it_up` part of the last test because we only care about this instant answer in this test file.

I kind of wanted to know what the "message" field in your server's response will be, since we'll be displaying that to the user. I'm available at https://gitter.im/mattr555 if you want to chat!

File: lib/DDG/Spice/MinecraftStatus.pm [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+primary_example_queries "is minecraft down", "minecraft status";
+category 'special';
+topics 'words_and_games';
+code_url "https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/MinecraftStatus.pm";
+attribution 
+	web     => ['https://thedestruc7i0n.ca', "Colin D'Souza"],
+	twitter => ['https://www.twitter.com/TheDestruc7i0n',"Colin D'Souza"],
+	github  => ['destruc7i0n', "Colin D'Souza"];
+
+triggers any => 'minecraft', 'mcstatus', 'mc';
```
Comment (jagtalon) [inline lib/DDG/Spice/MinecraftStatus.pm]: Maybe "mine craft" as well?

**Comment by jagtalon** [comment]
> It's cool seeing all of these new status IAs; maybe we could make a new template group for them?

@mattr555 Yes! Thanks for reminding me

**Comment by mattr555** [comment]
@jagtalon: @destruc7i0n and I talked and we said that if there are shaky or down services, they would be displayed in a comma-separated list below, with colors. Screenshots to come I suppose!

File: lib/DDG/Spice/MinecraftStatus.pm [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+spice to => 'http://direct.thedestruc7i0n.ca/mcstatusraw';
```
Comment (jagtalon) [inline lib/DDG/Spice/MinecraftStatus.pm]: Looks like this is redirecting (and we can't handle redirects when its deployed on a server). Looks like adding a `/` at the end works.

File: lib/DDG/Spice/MinecraftStatus.pm (hunk 3 above)
Comment (destruc7i0n) [inline lib/DDG/Spice/MinecraftStatus.pm]: Good idea! Will add :smile:

**Comment by destruc7i0n** [comment]
Added image of new design! :rocket:

**Comment by destruc7i0n** [comment]
Sorry pushed wrong button

**Comment by jagtalon** [comment]
@abeyang what do you think of this:



@destruc7i0n @mattr555 @abeyang I think the up, shaky, and down should only be shown depending on the overall status. For example, if it was up, nothing would be shown, but if it was shaky or down, I think that's the only time when this is shown because people would know which ones got affected.

What do you think?

**Comment by destruc7i0n** [comment]
@mattr555 @abeyang @jagtalon I was actually planning on commiting that within the next hour, think it would be better? I would think that people may get confused if there is **too** much information there, remember that the main people playing Minecraft are kids and a few adults.

**Comment by jagtalon** [comment]
@destruc7i0n yeah I agree--it should be simpler, although I would wait for @abeyang's comments. :)

**Comment by destruc7i0n** [comment]
@abeyang @jagtalon Ah ok, let me see what I can do! Also, it may be a little hard to prioritize the byline, for example, if the Textures and Skins are shaky and the Session is completely down, how would I prioritize them? Would I put the ones that are shaky in the byline or the only 1 that is in the down category? I was actually going to do it this way, but this was the problem that made me change my idea.

**Comment by destruc7i0n** [comment]
Ah, I fixed things up on my side and got a commit ready, will commit in a few hours.
Image

**Comment by abeyang** [comment]
Hi @destruc7i0n, thanks for the quick turnaround! Also, you raised some good questions.

For your first image, it seems like Minecraft is experiencing **major** issues? But you have "minor" written there.

Let's try this byline for the various issues -- right now, we're using redundant terms (minor/minor and major/major):
* Minor issues: X is partially degraded.
* Major issues: X is down. (Unless a "major" issue is not necessarily down?)

To answer your question, I think the most severe should surface. So, if X is major and Y is minor, we should have:

> **Minecraft is experiencing _major_ issues**
> X is down; Y is partially degraded.

What do you think?

**Comment by destruc7i0n** [comment]
@abeyang 
First image: The reason was that it was the **ONLY** service down, it would make people think that the whole Minecraft is down, that is not really true as only one is.

Also, I will make sure to add those bylines!

**Comment by destruc7i0n** [comment]
@abeyang I am now actually thinking about removing the ones that are up because there are mainly more of them... But what if one day almost all go down and people want to know what is up? Is there a formula other IA's use for this type of thing or is this the only one?

**Comment by abeyang** [comment]
@destruc7i0n I don't think we need to mention which ones are up. Also, in your screenshot, we don't need to show that "none are partially degraded."

Even if one is down, I'd warrant that Minecraft is experiencing **major** issues. This is the standard that services like statuspage.io (and github, I believe) has taken, and I think it makes sense. Sooo, according to your screenshot, it should look like this:

> **Minecraft is experiencing _major_ issues**
> Realms is down.

(BTW, I'm working under the assumption that Minecraft users will understand "realms", "session", "skins," etc...)

**Comment by jagtalon** [comment]
@destruc7i0n I think I would like to trigger this with "status of minecraft". @moollaza what's your opinion on triggering this on just "minecraft"? I think it would be a pleasant surprise when people see that as one of the tabs:

screen shot 2015-01-29 at 2 14 26 pm

**Comment by jagtalon** [comment]
Now that I think about it, probably "github" and "travisci" should trigger their respective status tabs, eh?

**Comment by destruc7i0n** [comment]
@jagtalon That would be nice, but then all the regex would and I am still pretty new to that (and Perl), if someone could help that would be nice.
@abeyang Ok, I can make those changed you asked for, coming right up :)

**Comment by destruc7i0n** [comment]
@abeyang Did what you asked, here are the results:
Everything is up: Image
Major: Image
Shaky: Image

There are a few other combinations but I had enough image uploading :stuck_out_tongue_closed_eyes: :
If you want anymore images, just ask :wink:

**Comment by abeyang** [comment]
Looks good! :+1:

**Comment by destruc7i0n** [comment]
Thanks, now just to figure out the regex, if anyone of you know regex well so I can do what @jagtalon said earlier:

> @destruc7i0n I think I would like to trigger this with "status of minecraft". @moollaza what's your opinion on triggering this on just "minecraft"? I think it would be a pleasant surprise when people see that as one of the tabs:

If anyone could help that would be nice! :wink:

**Comment by moollaza** [comment]
@jagtalon I think we should leave the triggers as they are for now, and then expose a fn that can generate these kinds of "status" triggers so we can use it in all the other IA's that check the status of something. We have quite a few now and the triggering for all of them might as well be identical.

**Comment by destruc7i0n** [comment]
Judging by this the regex is not needed!
Is this ready to go or is there still some more steps in the process? (Sorry for rushing)

**Comment by jagtalon** [comment]
@destruc7i0n sweet! We'll review your code and get back to you. :)

**Comment by destruc7i0n** [comment]
@jagtalon Thanks! Waiting for the response, no rush :)

**Comment by destruc7i0n** [comment]
I just updated some code and made it a bit faster on my side, I also had the wrong server for Session so it would always be down. :stuck_out_tongue: 
Also is there any updates on this?

**Comment by moollaza** [comment]
@destruc7i0n sorry about the delay, we'll give this a look shortly!

**Comment by moollaza** [comment]
Everything LGTM :+1: 

Thanks a lot @destruc7i0n

**Comment by jagtalon** [comment]
Awesome!

**Comment by destruc7i0n** [comment]
Sweet! :smiley:

**Comment by moollaza** [comment]
@destruc7i0n congrats! It's live: https://duckduckgo.com/?q=minecraft+status&ia=status

**Pull request merged**